### DISCUSSION TITLE
Feedback on a script

### DISCUSSION THREAD

**Discussion by PPetterssonSO** [SuperOffice/typescript-for-superoffice#23] [General]
Hey hey!
I might ask a few "What's the best practice" here, since the code library im building now will be in MANY solutions after TypeScript goes live, so, better safe then sorry (Sorry = inefficient code that will haunt your azure bills)

So, since we are about 16 developers building SearchEngines multiple times a day, that seems to be a perfect opportunity to streamline a bit in typescript.
(And yes, In this case, im using a getArchiveListByColumns2Async, will build another one for the SimpleContac etct)

 Can you see any approvements to this code when it comes to  efficiency?
################

function printLine(value){
 	context.result.body += value + "\r\n";
};

const agent = soApi.getArchiveAgent();
async function searchEngine(fields : Set<string>, criteria : string, sortOrder? : string,  maxRows? : number)
{
 const fieldsData: string[] = Array.from(fields);
 const fieldString: string = Array.from(addFields).join(",");
 let rowLimit = 999;
 if(maxRows !== undefined)
   rowLimit = maxRows;
  
    let result = await agent.getArchiveListByColumns2Async(
    "dynamic",
    fieldString, 
    sortOrder,
    criteria,
    "",
    0,
    rowLimit
    );
    
var response = result.map(item => {
    const extractedData: any = {};
    fieldsData.forEach(field => {
      if (item.columnData[field]) {
        extractedData[field] = item.columnData[field].displayValue;
      }
    });
    return extractedData;
  });
 return response;
}

const addFields = new Set<string>();
addFields.add("contact.contact_id");
addFields.add("contact.name");
addFields.add("contact.category_idx");
var searchEngineResults = await searchEngine(addFields,"contact.contact_id gt 5 AND contact.contact_id lt 15");

searchEngineResults.forEach(item => {
  printLine(item["contact.contact_id"]);
});

################

Also, is there any documentation or examples on how to handle Priorities in the getArchiveAgent?
getCompanies.addCriteria("contact.contact_id","Equals","2","And", <THIS NUMBER>);

**Comment by ejfasting** [comment]
Hello @Nawksen, 

Very exciting that you are trying things out and eager to get started!

We are still relatively early in the development and i `think` what you are asking for is something we can assist with a little further down the line. 
The benefit of writing TypeScript is that it is now in line with what you can do in the REST API, so it will be easier to both understand what is possible AND follow the same practices as you would while using the [@superoffice/webapi](https://www.npmjs.com/package/@superoffice/webapi?activeTab=code) package. 

###

I would suggest you take a look at the [dynamic archive provider](https://docs.superoffice.com/en/api/netserver/search/odata/dynamic-provider.html) and how you can create queries with the dot-syntax similar to how you would with the SearchEngine. 

[Priorities](https://docs.superoffice.com/en/automation/crmscript/searchengine/se-select.html#where) are just a way to group criterias in the resulting SQL query, so you should be able to do the same with the dot-syntax. 

If you need anything more specific then I am more then happy to assist if you have a concrete example I can `transform` for you :) 

//E
